stdlib/string.ml

```diff
@@ -225,17 +226,18 @@ let ends_with ~suffix s =
 external seeded_hash : int -> string -> int = "caml_string_hash" [@@noalloc]
 let hash x = seeded_hash 0 x
 
-(* duplicated in bytes.ml *)
+(* duplicated in bytes.ml, except for the case returning [s] *)
 let split_on_char sep s =
   let r = ref [] in
-  let j = ref (length s) in
-  for i = length s - 1 downto 0 do
+  let len = length s in
+  let j = ref len in
+  for i = len - 1 downto 0 do
     if unsafe_get s i = sep then begin
       r := sub s (i + 1) (!j - i - 1) :: !r;
       j := i
     end
   done;
-  sub s 0 !j :: !r
+  if !j = len then [s] else sub s 0 !j :: !r
```

Comment:
Have you considered optimizing String.sub instead ?

Reply:
No I did not consider this! Good idea!

Reply:
I did this, removed the special case for split_on_char and optimized String.sub. The comment is back as it is now a real duplicate.